components/layout/inline.rs

```diff
@@ -919,18 +958,47 @@ impl InlineFlow {
             text_align::T::left | text_align::T::right => unreachable!()
         }
 
-        for fragment_index in line.range.each_index() {
-            let fragment = fragments.get_mut(fragment_index.to_usize());
-            inline_start_position_for_fragment = inline_start_position_for_fragment +
-                fragment.margin.inline_start;
-            fragment.border_box = LogicalRect::new(fragment.style.writing_mode,
-                                                   inline_start_position_for_fragment,
-                                                   fragment.border_box.start.b,
-                                                   fragment.border_box.size.inline,
-                                                   fragment.border_box.size.block);
-            fragment.update_late_computed_inline_position_if_necessary();
-            inline_start_position_for_fragment = inline_start_position_for_fragment +
-                fragment.border_box.size.inline + fragment.margin.inline_end;
+        // Lay out the fragments in visual order.
+        let run_count = match line.visual_runs {
+            Some(ref runs) => runs.len(),
+            None => 1
+        };
+        for run_idx in 0..run_count {
+            let (range, level) = match line.visual_runs {
+                Some(ref runs) if is_ltr => runs[run_idx],
+                Some(ref runs) => runs[run_count - run_idx - 1], // reverse order for RTL runs
+                None => (line.range, 0)
+            };
+            // If the bidi embedding direction is opposite the layout direction, lay out this
+            // run in reverse order.
+            let reverse = unicode_bidi::is_ltr(level) != is_ltr;
+            let (begin, end, step) = if reverse {
+                (range.end().get() - 1, range.begin().get() - 1, -1)
+            } else {
+                (range.begin().get(), range.end().get(), 1)
+            };
```

Comment:
I wonder if you can just create the iterator inside the then and else branches instead of creating all these temporaries.